hypothesis-python/src/hypothesis/stateful.py

```diff
@@ -103,20 +104,46 @@ def run_state_machine(factory, data):
             check_type(RuleBasedStateMachine, machine, "state_machine_factory()")
         data.conjecture_data.hypothesis_runner = machine
 
-        n_steps = settings.stateful_step_count
-        should_continue = cu.many(
-            data.conjecture_data, min_size=1, max_size=n_steps, average_size=n_steps
-        )
-
         print_steps = (
             current_build_context().is_final or current_verbosity() >= Verbosity.debug
         )
         try:
             if print_steps:
                 machine.print_start()
             machine.check_invariants()
+            max_steps = settings.stateful_step_count
+            steps_run = 0
+
+            cd = data.conjecture_data
+
+            while True:
+                # We basically always want to run the maximum number of steps,
+                # but need to leave a small probability of terminating early
+                # in order to allow for reducing the number of steps once we
+                # find a failing test case, so we stop with probability of
+                # 2 ** -16 during normal operation but force a stop when we've
+                # generated enough steps.
+                cd.start_example(STATE_MACHINE_RUN_LABEL)
+                if steps_run == 0:
+                    cd.draw_bits(16, forced=1)
+                elif steps_run >= max_steps:
+                    cd.draw_bits(16, forced=0)
```

Comment:
The `stateful_step_count` setting currently allows a value of `0` - it looks like this should be deprecated and the true minimum is one?

Reply:
In fact we can just turn it into a validation error withotu deprecation, because setting it to `0` would have previously triggered an internal assertion error, whoops!

I'll do that before this is merged so we don't go through a period where it silently does the wrong thing instead.